Make provided/implementedBy and adapter registries respect super().

jamadden opened the pull request:
The query functions now start by looking at the next class in the MRO (interfaces directly provided by the underlying object are not found).

Adapter registries automatically pick up providedBy change to start finding the correct implementations of adapters, but to make that really useful they needed to change to unpack super() arguments and pass `__self__` to the factory.

Fixes #11

Unfortunately, this makes PyPy unable to build the C extensions.

icemac commented:
@d-maurer I added you as reviewer as #11 was originally written by you.

d-maurer commented on CHANGES.rst:
```diff
@@ -134,6 +134,20 @@
+  Along with this change, adapter registries will unpack ``super``
+  objects into their ``__self___`` before passing it to the factory.
+  Together, this means that ``component.getAdapter(super(Derived,
+  self), ITarget)`` is now meaningful.
```
While the unpacking is sufficient for my original use case, this may not always be the case, i.e. in some cases, one might want that the super object (and not its `__self__`) is passed to the adapter. I agree that given the current interface, it is impossible to express what the use case wants. In the long term, I would like the interface to be enhanced by a `generateAdapterFactories` decoupling adapter factory lookup and calling them. This would also significantly facilitate the implementation of "conditional adapters" (a concept, I have used for `Products.AdvancedQeury`).

jamadden replied:
I don't have a strong immediate use case, but it seemed to me that passing the `super` object was both not very useful and inconsistent with other uses of both `super`. The practical uses I could think of for today all work best with `__self__` unwrapped.

It's inconsistent because when `super(Class, self).method()` is executing, the `self` that's sent to `method` isn't the `super` object, it's straight up exactly the `self` object provided to `super` (because `super` isn't particularly useful; see next paragraph).  Adapting on `super` ought to implicitly behave the same simply to match that expectation.

It's not particularly useful because the `super` object cannot access instance variables of `self` (only class descriptors/methods), nor (as a consequence) can it discover any interfaces that are `directlyProvidedBy` the `self` object. It has no MRO or type and `super` doesn't work on it, so you'd only be able to try this trick for exactly one level of hierarchy. The receiving factory would need to unwrap `__self__` in short order to do any of those things…and at that point, you're writing an adapter that's customized to know it's working with `super`, which destroys reusability: given `class Base` and `class Derived(Base)` the same adapter can't work for `super(Derived, self)` *and* `Base()`.

If one wanted to use the `super` object in an adapter, I think the better thing to do would be to register a multi-adapter on `(Interface-or-class, type)` (called with `component.getMultiAdapter(IAdapter, (self, Derived))`) or possibly `(Interface-or-class, super)`. The latter would work today (but not after this change), while the former would work today and after this change.

d-maurer commented on src/zope/interface/_zope_interface_coptimizations.c:
```diff
@@ -91,6 +95,41 @@ import_declarations(void)
+    mro = PyObject_CallMethodObjArgs(class_that_invoked_super, strmro, NULL);
+    _NSCHECK(mro, NULL);
+
+    index = PySequence_Index(mro, class_that_invoked_super);
```
I think that the `mro` must be determined from `ob.__self_class__` not from `ob.__thisclass__`. I expect that `index` above will always be `0`.

jamadden replied:
I'm afraid you have that backwards. In `super(Class, self)` `__self_class__` is always the same as `type(self)`. It never changes and is the most derived class in the MRO (index 0).  It's `__thisclass__` that changes to reflect where in the MRO the new starting search point actually is.

```pycon
>>> class A(object):
...     pass
...
>>> class B(A):
...     pass
...
>>> class C(B):
...     pass
...
>>> class D(C):
...     pass
...
>>> D.__mro__
(<class '__main__.D'>,
 <class '__main__.C'>,
 <class '__main__.B'>,
 <class '__main__.A'>,
 <class 'object'>)
>>> d = D()
>>> sd = super(D, d)
>>> sc = super(C, d)
>>> sb = super(B, d)

>>> (D, sd.__self_class__, sc.__self_class__, sb.__self_class__)
(<class '__main__.D'>,
 <class '__main__.D'>,
 <class '__main__.D'>,
 <class '__main__.D'>)

>>> sd.__thisclass__
<class '__main__.D'>
>>> sc.__thisclass__
<class '__main__.C'>
>>> sb.__thisclass__
<class '__main__.B'>
```

jamadden replied:
>> I'm afraid you have that backwards. In `super(Class, self)` `__self_class__` is always the same as `type(self)`. It never changes and is the most derived class in the MRO (index 0). It's `__thisclass__` that changes to reflect where in the MRO the new starting search point actually is.

> I am aware of this. That's why I am (quite) sure that
> what you need is
```
> mro = PyObject_CallMethodObjArgs(__self_class__, strmro, NULL);
> _NSCHECK(mro, NULL);
>
> index = PySequence_Index(mro, class_that_invoked_super);
```

> With your code, `index` will always be 0 as for any class
> it is the 0's element of its "mro".

Essentially true. We're letting `super` do the harder work of figuring out exactly where in MRO we are. I thought that the class linearization order, which preserves local precedence order, would handle everything else correctly while minimizing our search time here.  But I forgot about diamonds!

Continuing the example from above it looks good:
```pycon
>>> from zope.interface.declarations import _next_super_class
>>> _next_super_class(sd) # super(D, d) means begin looking above D
<class '__main__.C'>
>>> _next_super_class(sc)
<class '__main__.B'>
>>> _next_super_class(sb)
<class '__main__.A'>
```

Introducing a diamond shows the inconsistency:
```pycon
>>> class BB(A):
...     pass
...
>>> class E(B, BB):
...     pass
...
>>> E.__mro__
(<class '__main__.E'>,
 <class '__main__.B'>,
 <class '__main__.BB'>,
 <class '__main__.A'>,
 <class 'object'>)
>>> e = E()
>>> se = super(E, e)
>>> sb = super(B, e)
>>> sbb = super(BB, e)
>>> _next_super_class(se)
<class '__main__.B'>
>>> _next_super_class(sb) # Should be BB
<class '__main__.A'>
>>> _next_super_class(sbb)
<class '__main__.A'>
```

Redefining as suggested is consistent. Thank you!
```pycon
>>> def _next_super_class(ob):
...     # When ``ob`` is an instance of ``super``, return
...     # the next class in the MRO that we should actually be
...     # looking at.
...     class_that_invoked_super = ob.__self_class__
...     mro = class_that_invoked_super.__mro__
...     next_class = mro[mro.index(ob.__thisclass__) + 1]
...     return next_class
…
>>> _next_super_class(se)
<class '__main__.B'>
>>> _next_super_class(sb)
<class '__main__.BB'>
>>> _next_super_class(sbb)
<class '__main__.A'>
```

d-maurer commented on src/zope/interface/declarations.py:
```diff
@@ -54,6 +54,15 @@ class implements (that instances of the class provides).
+    # looking at.
+    class_that_invoked_super = ob.__thisclass__
+    mro = class_that_invoked_super.mro()
+    next_class = mro[mro.index(class_that_invoked_super) + 1]
```
Again: the `mro` should be determined for `ob.__self_class_`, not `ob.__thisclass__`.

d-maurer commented:
Jason Madden wrote at 2020-3-7 01:44 -0800:
>jamadden commented on this pull request.
>
>
>
>> +    mro = PyObject_CallMethodObjArgs(class_that_invoked_super, strmro, NULL);
>+    _NSCHECK(mro, NULL);
>+
>+    index = PySequence_Index(mro, class_that_invoked_super);
>
>I'm afraid you have that backwards. In `super(Class, self)` `__self_class__` is always the same as `type(self)`. It never changes and is the most derived class in the MRO (index 0).  It's `__thisclass__` that changes to reflect where in the MRO the new starting search point actually is.

I am aware of this. That's why I am (quite) sure that
what you need is

    mro = PyObject_CallMethodObjArgs(__self_class__, strmro, NULL);
    _NSCHECK(mro, NULL);

    index = PySequence_Index(mro, class_that_invoked_super);

With your code, `index` will always be 0 as for any class
it is the 0's element of its "mro".

d-maurer commented:
Jason Madden wrote at 2020-3-7 02:05 -0800:
> ...
>It's not particularly useful because the `super` object cannot access instance variables of `self` (only class descriptors/methods), nor (as a consequence) can it discover any interfaces that are `directlyProvidedBy` the `self` object. It has no MRO or type and `super` doesn't work on it, so you'd only be able to try this trick for exactly one level of hierarchy. The receiving factory would need to unwrap `__self__` in short order to do any of those things��and at that point, you're writing an adapter that's customized to know it's working with `super`, which destroys reusability: given `class Base` and `class Derived(Base)` the same adapter can't work for `super(Derived, self)` *and* `Base()`.


I am convinced.

jamadden commented:
CPython 3.5 is crashing on Travis/Linux, apparently hanging on Travis/Mac, and crashing locally for me on my Mac. 

The local crash occurs in `test_getMultiAdapter_hit_super (zope.interface.tests.test_registry.ComponentsTests)` and looks like an issue computing a hash code (?) when getting a key in a dictionary (compiled with `-g -O0`:

```bt
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00007fff67191774 libsystem_malloc.dylib`large_malloc + 89
    frame #1: 0x00007fff67187357 libsystem_malloc.dylib`szone_malloc_should_clear + 235
    frame #2: 0x00007fff6718606a libsystem_malloc.dylib`malloc_zone_malloc + 104
    frame #3: 0x00007fff67185fe5 libsystem_malloc.dylib`malloc + 21
    frame #4: 0x000000010012f012 .Python`_PyObject_Alloc + 881
    frame #5: 0x0000000100120030 .Python`PyLong_FromDouble + 176
    frame #6: 0x00000001026f1f23 _zope_interface_coptimizations.cpython-35m-darwin.so`_lookup(self=0x000000010631cea8, required=0x0000000106322b08, provided=0x00000001060ff9a0, name=0x0000000100553ab0, default_=0x0000000000000000) at _zope_interface_coptimizations.c:1015:12
    frame #7: 0x00000001026f1b59 _zope_interface_coptimizations.cpython-35m-darwin.so`lookup_lookup(self=0x000000010631cea8, args=0x00000001063195e8, kwds=0x0000000000000000) at _zope_interface_coptimizations.c:1060:10
    frame #8: 0x000000010012c1f7 .Python`PyCFunction_Call + 107
```

where `_zope_interface_coptimizations.c:1015` is `result = PyDict_GetItem(cache, key);`

jamadden commented:
The crash in 3.5 turned out to be a memory bug in ``tuple(map(func, it))`` when `func` could return temporary objects. The fix is to replace `map(func, it)` with `[func(i) for i in it]`. This turns out to be faster too.

The temporary objects in question are the objects created for `providedBy(super(C, c))`. Unlike `providedBy(c)`, those aren't currently cached anywhere else (in an instance or class dict). But they'll wind up cached somewhat invisibly in adapter registries. They'll be found again (hashcodes and eq match up) so we're just doing some extra work computing the RO each time. I'm wondering if we should cache that in the class dict to avoid that.

Right now, I'm thinking leave it simple and revisit if it becomes a bottleneck.

jamadden commented:
> I am also worried that the use of tuple(map(func, it)) (over [func(i) for i in it]) could cause a crash. In my view, this indicates that something is deeply wrong and may hit people without the capabilities to find out why and how to work around it. Did the problem only affected Python 3.5 or other Python 3 versions as well (in other works, were the bug in Python 3.5 or somewhere in zope.interface).

Yeah, I was concerned about that too and kept looking into it. The crash only happened in CPython 3.5, and debugging showed that the call to `tuplefy(required)` resulted in the  dictionary stored in the `cache` C variable getting deallocated. Since `tuplefy` was iterating a lazy mapping that's calling arbitrary Python code, arbitrary side-effects are possible. In this case, it turns out that one such side-effect was eventually calling `components.changed(_)`, which drops the caches. Since `cache` contained a borrowed reference, its refcount went to 0 and it got deallocated. (This happened on every version of Python, but for some reason having to do with the `dict` implementation or memory management, it only crashed on CPython 3.5.)

The change to non-lazily iterate before calling into C is a sufficient fix *for this code* (although I can and will also eliminate the invoking of this side-effect in the first place). But this is a bug waiting to happen and the C code has been incorrect for a long time. The remainder of the fix is to not get the cache until we can be (more) sure there will be no more arbitrary side effects.

jamadden commented:
> Something is still inconsistent….Either tuple(providedBy(super(M1, d))) should not contain IB or tuple(providedBy(super(M2, d))) should contain IB (because M1 and M2 are symmetric with respect to B). Very likely, the latter should be the case, as super(M2, d) is "d viewed as B object" and B implements IB.

You're absolutely right. I've added that as a test case and corrected it.

d-maurer commented on src/zope/interface/adapter.py:
```diff
@@ -30,6 +30,22 @@
+# ``tuple(t for t in range(10))``        -> 665ns
+# ``tuple(map(lambda t: t, range(10)))`` -> 958ns
+#
+# All three have substantial variance.
```
The comment looks somewhat unmotivated at this place (though I know from the conversation in this PR why it is somewhere).

jamadden commented:
Thanks for the detailed reviews and attention to detail!

> Should we have a test "adapter_hook hits super" to verify that the "__self__" dereferencing works in this case, too? Maybe, also for multi adapters?

Both of those things are covered in the tests, but I will look at making them more explicit.
